src/engine/engine/Application.cpp

```diff
@@ -0,0 +1,208 @@
+        glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE);
+
+        #ifdef __APPLE__
+            glfwWindowHint(GLFW_OPENGL_FORWARD_COMPAT, GL_TRUE);
+        #endif
+
+        int screenWidth = 1200;
+        int screenHeight = 1000;
+
+        m_Window = glfwCreateWindow(screenWidth, screenHeight, "lei3d", NULL, NULL);
+        if (m_Window == NULL)
+        {
+            LEI_WARN("failed to create GLFW window");
+            glfwTerminate();
+            return;
+        }
+        glfwMakeContextCurrent(m_Window);
+
+        if (!gladLoadGLLoader((GLADloadproc)glfwGetProcAddress))
+        {
+            LEI_WARN("failed to initialize GLAD");
+            return;
+        }
+
+        GLCall(glViewport(0, 0, screenWidth, screenHeight));
+
+        // resize the openGL context if a user changes the window size
+        glfwSetFramebufferSizeCallback(m_Window, framebuffer_size_callback);
+
+        // set up imgui
+        ImGui::CreateContext();
+        ImGui_ImplGlfw_InitForOpenGL(m_Window, true);
+        ImGui_ImplOpenGL3_Init();
+    }
+
+    void Application::LoadScene(Scene* scene)
+    {
+        GLCall(glEnable(GL_DEPTH_TEST));
+
+        m_ActiveScene = scene;
+        m_ActiveScene->Init(this);
+        SetupSceneCallbacks();
+    }
+
+    void Application::SetupSceneCallbacks() {
+        glfwSetWindowUserPointer(m_Window, this);
+        glfwSetInputMode(m_Window, GLFW_CURSOR, GLFW_CURSOR_DISABLED); // turn off if UI
+        glfwSetCursorPosCallback(m_Window, [](GLFWwindow* window, double x, double y)
+            {
+                Application* self = static_cast<Application*>(glfwGetWindowUserPointer(window));
+                if (self)
+                {
+                    if (glfwGetInputMode(window, GLFW_CURSOR) == GLFW_CURSOR_DISABLED) // when the cursor is normal, don't move the camera according to mouse mvnt
+                        self->m_ActiveScene->MainCamera()->cameraMouseCallback(window, x, y);
+                }
+            });
+
+        glfwSetKeyCallback(m_Window, [](GLFWwindow* window, int key, int scancode, int action, int mods)
+            {
+                Application* self = static_cast<Application*>(glfwGetWindowUserPointer(window));
+                if (self)
+                {
+                    self->processInput(window, key, scancode, action, mods);
+                }
+            });
+    }
+
+    void Application::FrameTick() {
+        const float currentTime = (float)glfwGetTime();
+
+        deltaTime = currentTime - lastFrameTime;
+        lastFrameTime = currentTime;
+
+        glfwPollEvents();
+        Update();
+        Render();
+
+        m_ActiveScene->ProcessCameraInput(deltaTime);
+        glfwSwapBuffers(m_Window);
+
+        //Sync FPS to desired value
+        const float currentDrawTime = (float)glfwGetTime() - currentTime;
+        const float sleepTime = 1.0f / desiredFPS - currentDrawTime;
+        if (sleepTime > 0.0f) {
+            _sleep(sleepTime);
```

Comment:
make sure this doesn't break on Linux.

Note from yourself: "As long as you have access to C++11 or higher I would use std::this_thread::sleep_for and let the standard library worry about calling the correct "sleep" function."